shader/flatten.wgsl

```diff
@@ -333,33 +383,57 @@ fn read_path_segment(tag: PathTagData, transform: Transform, is_stroke: bool) ->
         // This is encoded this way because encoding this as a lineto would require adding a moveto,
         // which would terminate the subpath too early (by setting the SUBPATH_END_BIT on the
         // segment preceding the cap marker). This scheme is only used for strokes.
-        p0 = transform_apply(transform, p1);
-        p1 = transform_apply(transform, p2);
+        p0 = p1;
+        p1 = p2;
         seg_type = PATH_TAG_LINETO;
-    } else {
-        p0 = transform_apply(transform, p0);
-        p1 = transform_apply(transform, p1);
     }
 
     // Degree-raise
     if seg_type == PATH_TAG_LINETO {
         p3 = p1;
         p2 = mix(p3, p0, 1.0 / 3.0);
         p1 = mix(p0, p3, 1.0 / 3.0);
-    } else if seg_type >= PATH_TAG_QUADTO {
-        p2 = transform_apply(transform, p2);
-        if seg_type == PATH_TAG_CUBICTO {
-            p3 = transform_apply(transform, p3);
-        } else {
-            p3 = p2;
-            p2 = mix(p1, p2, 1.0 / 3.0);
-            p1 = mix(p1, p0, 1.0 / 3.0);
-        }
+    } else if seg_type == PATH_TAG_QUADTO {
+        p3 = p2;
+        p2 = mix(p1, p2, 1.0 / 3.0);
+        p1 = mix(p1, p0, 1.0 / 3.0);
     }
 
     return CubicPoints(p0, p1, p2, p3);
 }
 
+fn output_line(path_ix: u32, p0: vec2f, p1: vec2f) {
+    let line_ix = atomicAdd(&bump.lines, 1u);
+    bbox = vec4(min(bbox.xy, min(p0, p1)), max(bbox.zw, max(p0, p1)));
+    lines[line_ix] = LineSoup(path_ix, p0, p1);
+}
+
+fn output_line_with_transform(path_ix: u32, p0: vec2f, p1: vec2f, transform: Transform) {
+    let line_ix = atomicAdd(&bump.lines, 1u);
+    let tp0 = transform_apply(transform, p0);
+    let tp1 = transform_apply(transform, p1);
+    bbox = vec4(min(bbox.xy, min(tp0, tp1)), max(bbox.zw, max(tp0, tp1)));
+    lines[line_ix] = LineSoup(path_ix, tp0, tp1);
+}
+
+fn output_two_lines_with_transform(
+    path_ix: u32,
+    p00: vec2f, p01: vec2f,
+    p10: vec2f, p11: vec2f,
+    transform: Transform
+) {
+    let line_ix = atomicAdd(&bump.lines, 2u);
```

Comment:
Is this semantically equivalent to `output_line_with_transform` called twice, and only written this way to reduce atomic traffic?

Reply:
Thanks, that makes sense.

I'd probably add a comment to that effect, or possible add a function (`output_line_into_allocated_space`?) which these both call

I did notice one case where four lines were handled in a row (when handling miter joins) . Handling that case all in one bump could make sense (and could additionally avoid creating some `var`iables). But that also adds would seem to add quite a lot of noise for tiny gains

Reply:
I think this is a good idea. I'm not sure yet how significant the impact of the additional atomic here is but I suspect it might matter. At any rate, I like the idea of adding a `output_line_into_allocated_space` helper as I need something to that effect for round caps and joins where the line count is typically a lot higher.